File: python/binary-search/median-of-two-sorted-arrays.py
class Solution:
    def findMedianSortedArrays(self, nums1: List[int], nums2: List[int]) -> float:
        k = (len(nums1) + len(nums2))
        odd = False
        if k % 2 == 1:
            k = k // 2
            k += 1
            odd = True
        else:
            k = k // 2

        while k > 1:
            if len(nums1) == 0 or len(nums2) == 0:
                break
            if k % 2 == 0:
                current = k // 2 - 1
                if current > len(nums1) - 1 or current > len(nums2) - 1:
                    current = min(len(nums1) - 1, len(nums2) - 1)
                num1 = nums1[current]
                num2 = nums2[current]
                if num1 > num2:
                    nums2 = nums2[current + 1:]
                else:
                    nums1 = nums1[current + 1:]
                k -= current + 1
            else:
                tempk = k - 1
                current = tempk // 2 - 1
                if current > len(nums1) - 1 or current > len(nums2) - 1:
                    current = min(len(nums1) - 1, len(nums2) - 1)
                num1 = nums1[current]
                num2 = nums2[current]
                if num1 > num2:
                    nums2 = nums2[current + 1:]
                else:
                    nums1 = nums1[current + 1:]
                k -= current + 1

        if len(nums1) == 0:
            if odd:
                return nums2[k - 1]
            else:
                return (nums2[k - 1] + nums2[k]) / 2
        if len(nums2) == 0:
            if odd:
                return nums1[k - 1]
            else:
                return (nums1[k - 1] + nums1[k]) / 2

        if odd:
            return min(nums1[0], nums2[0])
        else:
            one = nums1[0]
            two = nums2[0]
            if len(nums2) > 1 and nums2[1] < one:
                return (nums2[0] + nums2[1]) / 2
            elif len(nums1) > 1 and nums1[1] < two:
                return (nums1[0] + nums1[1]) / 2
            return (nums1[0] + nums2[0]) / 2

File: c++/binary-search/median-of-two-sorted-arrays.cc
class Solution {
public:
    double findMedianSortedArrays(vector<int>& nums1, vector<int>& nums2) {
        if (nums1.size() > nums2.size()) {
            return findMedianSortedArrays(nums2, nums1);
        }

        int m = nums1.size();
        int n = nums2.size();

        int left = 0;
        int right = m;

        while (left <= right) {
            int middle_1 = (left + right) / 2;
            int middle_2 = (m + n + 1) / 2 - middle_1;

            int max_left_1 = (middle_1 == 0) ? std::numeric_limits<int>::min() : nums1[middle_1 - 1];
            int min_right_1 = (middle_1 == m) ? std::numeric_limits<int>::max() : nums1[middle_1];
            int max_left_2 = (middle_2 == 0) ? std::numeric_limits<int>::min() : nums2[middle_2 - 1];
            int min_right_2 = (middle_2 == n) ? std::numeric_limits<int>::max() : nums2[middle_2];

            if (max_left_1 <= min_right_2 && max_left_2 <= min_right_1) {
                if ((m + n) % 2 == 0) {
                    return (std::max(max_left_1, max_left_2) + std::min(min_right_1, min_right_2)) / 2.0;
                } else {
                    return max(max_left_1, max_left_2);
                }
            }

            if (max_left_1 > max_left_2) {
                right = middle_1 - 1;
            } else {
                left = middle_1 + 1;
            }
        }
        return 0.0;
    }
};
